src/jbmc/jbmc_parse_options.cpp

```diff
@@ -571,28 +574,41 @@ int jbmc_parse_optionst::get_goto_program(
 
   try
   {
-    goto_model=initialize_goto_model(cmdline, get_message_handler());
+    lazy_goto_modelt lazy_goto_model(get_message_handler());
+    lazy_goto_model.initialize(cmdline);
+
+    status() << "Generating GOTO Program" << messaget::eom;
+    lazy_goto_model.load_all_functions();
 
+    // Show the symbol table before process_goto_functions mangles return
+    // values, etc
     if(cmdline.isset("show-symbol-table"))
     {
-      show_symbol_table(goto_model, ui_message_handler.get_ui());
+      show_symbol_table(
+        lazy_goto_model.symbol_table, ui_message_handler.get_ui());
       return 0;
     }
 
-    if(process_goto_program(options))
+    // Move the model out of the local lazy_goto_model
+    // and into the caller's goto_model
+    goto_model=lazy_goto_modelt::freeze(std::move(lazy_goto_model));
+    if(goto_model == nullptr)
+      return 6;
```

Comment:
This is only reproducing what would have happened before (passing on the failure of `process_goto_module`). I'd hope that all such paths would already have reported the actual error condition. Would you like this specific message to be followed by a summary (something like "Conversion of lazy goto model to concrete goto model failed. Exiting.")?

Reply:
Hmm, maybe I'm misunderstanding this here, but I was assuming that the failure of `freeze` is a very peculiar kind of bug. Is it really that a lot more could be going wrong in this step?

Reply:
Freeze calls `process_goto_functions` (a replacement for the non-function-specific parts of `process_goto_module`) which is where most causes of failure come from (the exception being a failure in the call to `final`, which has its own error message).

Reply:
Ok, that certainly invalidates my initial concern. But it seems the naming in this API is not particularly intuitive. My preferred search engine tells me that "freeze" means "store (something) at a very low temperature in order to preserve it" or "an act of holding or being held at a fixed level or in a fixed state." Yet you are saying that really quite a lot happens before we might enter such a state of preservation.

Reply:
Those definitions aren't quite right. To freeze something means to take it from it's previous state and bring it to the state you describe. In this case we are taking a goto_model that doesn't have a fixed set of methods yet and freezing it to create one where methods can no longer be loaded. Taking a read-write object and creating a read-only object from it is a fairly common use of the word freeze in programming.

Reply:
I had just taken those definitions from Google's Dictionary, so either Google is wrong or I misinterpreted something.

Anyway, I do agree with "taking a read-write object and creating a read-only object" -- but what one does _not_ know is what's going to happen in this process. Function identifiers can be a lot longer than 6 characters, and I'd strongly solicit clarity via descriptive names rather than obfuscation.

Reply:
How about simply `run_module_passes_and_freeze`, since the method (a) runs the module-level transform passes and then (b) gives you a plain-old `goto_modelt` back, two steps which must take place together since the module-level passes assume function loading is complete.

Reply:
Sounds good to me!

Reply:
Could @smowton's suggestion be followed, please?

Reply:
module means something else in the Java context so after discussion with Chris I've chosen a different name.